### ISSUE TITLE
三通道彩图的隐写分析

### ISSUE THREAD

**Issue by yhySerendipper** [brijeshiitg/XuNet-Structural-Design-of-Convolutional-Neural-Networksfor-Steganalysis#3]
我用于三通道彩图的隐写分析检测，但是我的损失函数一直保持在0.69左右，怎么解决呢

**Comment by yueg-security** [comment]
I used to have the same problem…

**Comment by rebornone1** [comment]
你们运行成功的torch版本是多少呢

**Comment by zengchen233** [comment]
> 我用于三通道彩图的隐写分析检测，但是我的损失函数一直保持在0.69左右，怎么解决呢

你好，我的也是，一直都保持在0.7左右，下不去也上不来。请问您是如何修改让它可以判别RGB图像的呢？
```python
class ImageProcessing(nn.Module):
    """Computes convolution with KV filter over the input tensor for 3-channel images."""

    def __init__(self):
        """Constructor"""
        super(ImageProcessing, self).__init__()
        # 创建一个3通道的卷积核，每个通道对应一个5x5的索贝尔边缘检测核
        self.kv_filter = torch.tensor([[
            [-1.0, 2.0, -2.0, 2.0, -1.0],
            [2.0, -6.0, 8.0, -6.0, 2.0],
            [-2.0, 8.0, -12.0, 8.0, -2.0],
            [2.0, -6.0, 8.0, -6.0, 2.0],
            [-1.0, 2.0, -2.0, 2.0, -1.0],
        ]]).repeat(9, 1, 1).view(3, 3, 5, 5).cuda() / 12.0  # 重复3次以匹配3通道，原始的索贝尔边缘检测核的系数总和为12，除以12是为了使得核的输出在没有偏置的情况下也不会改变图像的整体亮度。
        # print("kv_filter:",self.kv_filter.shape)

    def forward(self, inp):
        """Returns tensor convolved with KV filter for each channel independently"""
        # 由于卷积核现在是3通道的，可以直接应用于三通道图像，不需要调整padding或stride
        # print("in shape:",inp.shape)
        return F.conv2d(inp, self.kv_filter, stride=1, padding=2)


class Xu_ConvBlock(nn.Module):
    """This class returns building block for XuNet class."""
    def __init__(
        self,
        in_channels: int,
        out_channels: int,
        kernel_size: int,
        activation: str = "relu",
        abs: str = False,
    ) -> None:
        super(Xu_ConvBlock, self).__init__()

        if kernel_size == 5:
            self.padding = 2
        else:
            self.padding = 0

        if activation == "tanh":
            self.activation = nn.Tanh()
        else:
            self.activation = nn.ReLU()

        self.abs = abs
        self.conv = nn.Conv2d(
            in_channels,
            out_channels,
            kernel_size,
            stride=1,
            padding=self.padding,
            bias=False,
        )
        self.batch_norm = nn.BatchNorm2d(out_channels)
        self.pool = nn.AvgPool2d(kernel_size=5, stride=2, padding=2)

    def forward(self, inp: Tensor) -> Tensor:
        """Returns conv->batch_norm."""
        if self.abs:
            return self.pool(
                self.activation(self.batch_norm(torch.abs(self.conv(inp))))
            )
        return self.pool(self.activation(self.batch_norm(self.conv(inp))))

class NetD(nn.Module):
    """Implementation of XuNet."""
    """This class returns XuNet model."""

    def __init__(self):
        super(NetD, self).__init__()
        self.layer1 = Xu_ConvBlock(
            3, 8, kernel_size=5, activation="tanh", abs=True
        )
        self.layer2 = Xu_ConvBlock(8, 16, kernel_size=5, activation="tanh")
        self.layer3 = Xu_ConvBlock(16, 32, kernel_size=1)
        self.layer4 = Xu_ConvBlock(32, 64, kernel_size=1)
        self.layer5 = Xu_ConvBlock(64, 128, kernel_size=1)
        self.gap = nn.AdaptiveAvgPool2d(output_size=1)
        self.fully_connected = nn.Sequential(
            nn.Linear(in_features=128, out_features=128),
            nn.Dropout(p=0.5),
            nn.ReLU(inplace=True),
            nn.Linear(in_features=128, out_features=2),
            nn.LogSoftmax(dim=1),
        )

    def forward(self, image):
        """Returns logit for the given tensor."""
        # print(image.shape)
        with torch.no_grad():
            # print("image shape:",image.shape)
            out = ImageProcessing()(image)
            # print("out shape:",out.shape)
        out = self.layer1(out)
        out = self.layer2(out)
        out = self.layer3(out)
        out = self.layer4(out)
        out = self.layer5(out)
        out = self.gap(out)
        out = out.view(out.size(0), -1)
        out = self.fully_connected(out)
        # print(out.shape)
        return out
```

**Comment by Li-179119** [comment]
> > 我用于三通道彩图的隐写分析检测，但是我的损失函数一直保持在0.69左右，怎么解决呢
> 
> 你好，我的也是，一直都保持在0.7左右，下不去也上不来。请问您是如何修改让它可以判别RGB图像的呢？
> 
> class ImageProcessing(nn.Module):
>     """Computes convolution with KV filter over the input tensor for 3-channel images."""
> 
>     def __init__(self):
>         """Constructor"""
>         super(ImageProcessing, self).__init__()
>         # 创建一个3通道的卷积核，每个通道对应一个5x5的索贝尔边缘检测核
>         self.kv_filter = torch.tensor([[
>             [-1.0, 2.0, -2.0, 2.0, -1.0],
>             [2.0, -6.0, 8.0, -6.0, 2.0],
>             [-2.0, 8.0, -12.0, 8.0, -2.0],
>             [2.0, -6.0, 8.0, -6.0, 2.0],
>             [-1.0, 2.0, -2.0, 2.0, -1.0],
>         ]]).repeat(9, 1, 1).view(3, 3, 5, 5).cuda() / 12.0  # 重复3次以匹配3通道，原始的索贝尔边缘检测核的系数总和为12，除以12是为了使得核的输出在没有偏置的情况下也不会改变图像的整体亮度。
>         # print("kv_filter:",self.kv_filter.shape)
> 
>     def forward(self, inp):
>         """Returns tensor convolved with KV filter for each channel independently"""
>         # 由于卷积核现在是3通道的，可以直接应用于三通道图像，不需要调整padding或stride
>         # print("in shape:",inp.shape)
>         return F.conv2d(inp, self.kv_filter, stride=1, padding=2)
> 
> 
> class Xu_ConvBlock(nn.Module):
>     """This class returns building block for XuNet class."""
>     def __init__(
>         self,
>         in_channels: int,
>         out_channels: int,
>         kernel_size: int,
>         activation: str = "relu",
>         abs: str = False,
>     ) -> None:
>         super(Xu_ConvBlock, self).__init__()
> 
>         if kernel_size == 5:
>             self.padding = 2
>         else:
>             self.padding = 0
> 
>         if activation == "tanh":
>             self.activation = nn.Tanh()
>         else:
>             self.activation = nn.ReLU()
> 
>         self.abs = abs
>         self.conv = nn.Conv2d(
>             in_channels,
>             out_channels,
>             kernel_size,
>             stride=1,
>             padding=self.padding,
>             bias=False,
>         )
>         self.batch_norm = nn.BatchNorm2d(out_channels)
>         self.pool = nn.AvgPool2d(kernel_size=5, stride=2, padding=2)
> 
>     def forward(self, inp: Tensor) -> Tensor:
>         """Returns conv->batch_norm."""
>         if self.abs:
>             return self.pool(
>                 self.activation(self.batch_norm(torch.abs(self.conv(inp))))
>             )
>         return self.pool(self.activation(self.batch_norm(self.conv(inp))))
> 
> class NetD(nn.Module):
>     """Implementation of XuNet."""
>     """This class returns XuNet model."""
> 
>     def __init__(self):
>         super(NetD, self).__init__()
>         self.layer1 = Xu_ConvBlock(
>             3, 8, kernel_size=5, activation="tanh", abs=True
>         )
>         self.layer2 = Xu_ConvBlock(8, 16, kernel_size=5, activation="tanh")
>         self.layer3 = Xu_ConvBlock(16, 32, kernel_size=1)
>         self.layer4 = Xu_ConvBlock(32, 64, kernel_size=1)
>         self.layer5 = Xu_ConvBlock(64, 128, kernel_size=1)
>         self.gap = nn.AdaptiveAvgPool2d(output_size=1)
>         self.fully_connected = nn.Sequential(
>             nn.Linear(in_features=128, out_features=128),
>             nn.Dropout(p=0.5),
>             nn.ReLU(inplace=True),
>             nn.Linear(in_features=128, out_features=2),
>             nn.LogSoftmax(dim=1),
>         )
> 
>     def forward(self, image):
>         """Returns logit for the given tensor."""
>         # print(image.shape)
>         with torch.no_grad():
>             # print("image shape:",image.shape)
>             out = ImageProcessing()(image)
>             # print("out shape:",out.shape)
>         out = self.layer1(out)
>         out = self.layer2(out)
>         out = self.layer3(out)
>         out = self.layer4(out)
>         out = self.layer5(out)
>         out = self.gap(out)
>         out = out.view(out.size(0), -1)
>         out = self.fully_connected(out)
>         # print(out.shape)
>         return out

请问解决了吗？
